onnx/test/version_converter_test.py

```diff
@@ -2184,6 +2186,78 @@ def test(input_shape, scale_shape, axis, block_size) -> None:
+
+            initializer_tensor = onnx.helper.make_tensor(
+                name="initializer_tensor",
+                data_type=onnx.TensorProto.FLOAT,
+                dims=list(shape),
+                vals=random_data.tobytes(),
+                raw=True,
+            )
+            initializer_scalar = onnx.helper.make_tensor(
+                name="initializer_scalar",
+                data_type=onnx.TensorProto.FLOAT,
+                dims=[],
+                vals=[1.0],
+            )
+
+            add_node = onnx.helper.make_node(
+                "Add",
+                inputs=["initializer_tensor", "initializer_scalar"],
+                outputs=["sum_output"],
+            )
+
+            graph_def = onnx.helper.make_graph(
+                name="SimpleAddition",
+                nodes=[add_node],
+                inputs=[],
+                outputs=[
+                    onnx.helper.make_tensor_value_info(
+                        "sum_output", onnx.TensorProto.FLOAT, list(shape)
+                    )
+                ],
+                initializer=[initializer_tensor, initializer_scalar],
+            )
+
+            # Save model to file with external data
+            model_filename = os.path.join(temp_dir, "test_simple_add.onnx")
+            data_filename = "test_simple_add.onnx.data"  # Use relative path
+            opset_imports = [onnx.helper.make_opsetid("", 20)]
+            model_def = onnx.helper.make_model(graph_def, opset_imports=opset_imports)
+            model_def.ir_version = 10
+            onnx.save_model(
+                model_def,
+                model_filename,
+                save_as_external_data=True,
+                all_tensors_to_one_file=True,
+                location=data_filename,
+                size_threshold=0,
+                convert_attribute=False,
+            )
+
+            # Load the model and verify external data
+            converted_model = onnx.version_converter.convert_version(
+                onnx.load(model_filename, load_external_data=False), 21
+            )
+            self.assertEqual(len(converted_model.graph.initializer), 2)
+
+            # Verify the large tensor has external data
+            for i in range(2):
+                if converted_model.graph.initializer[i].name == "initializer_tensor":
+                    large_tensor = converted_model.graph.initializer[i]
+                    self.assertEqual(len(large_tensor.external_data), 3)
+                    self.assertEqual(large_tensor.external_data[0].key, "location")
+                    self.assertEqual(large_tensor.external_data[0].value, data_filename)
+                    self.assertEqual(large_tensor.external_data[1].key, "offset")
+                    self.assertEqual(large_tensor.external_data[1].value, "0")
+                    self.assertEqual(large_tensor.external_data[2].key, "length")
+                    self.assertEqual(large_tensor.external_data[2].value, "24")
```

Comment:
The key-value pairs are not guaranteed in order. I would build a dictionary from `large_tensor.external_data` then assert the python dict directly.

Reply:
Got it. Updated.

Reply:
The test passes for me locally but in pipeline the failure seems like it's not picking up the source code changes in this PR. For example, it sees 0 external data and the data_location is not set to `TensorProto.EXTERNAL`. Do you know if it's using a cached version of ONNX to run the tests?

Reply:
The failed pipeline can be ignored. I am updating it in https://github.com/onnx/onnx/pull/6861. Thanks